### PR TITLE
Add an optimization to replace repeated array accesses with ref temps

### PR DESCRIPTION
**Pull request by daviditen**
Add a pragma "removable array access" and apply it to the array accessor (this)
functions in ChapelArray.

Within for loops, look for calls to functions that have the new pragma and
use the loop index variable as the index argument.  If more than a threshold
of these calls exist within a loop, assign one call to a reference variable and
replace the other calls with the reference.

For example:
```
  for i in rng {
    A[i] = A[i] * A[i];
    B[i] = A[i] + f(A[i]);
  }
```

Becomes:
```
  for i in rng {
    ref tmp = A[i];
    tmp = tmp * tmp;
    B[i] = tmp * f(tmp);
  }
```

This is controlled by the option --[no-]replace-array-accesses-with-ref-temps.
It is off by default and --fast and --baseline do not modify it at the moment.  It
is currently considered an "experimental" optimization and is opt-in.  Some
cases that are problematic are shown in the .futures in
test/arrays/diten/replaceArrayAccess/.

### CODE REVIEW

**Comment by daviditen** [comment]
Tested with a full paratest on linux64 and linux64+gasnet with "-compopts --replace-array-accesses-with-ref-temps".

**Comment by ronawho** [comment]
High level notes/questions without having looked at the code yet: 

- The compiler performance test (performance/compiler/compilerPerformance.perfkeys) and compiler performance .dat files will have to be updated (mostly a note to myself, nothing that you have to do.)

- Will this break if two different arrays alias each other? e.g.: 

  ```chapel

  proc someProc(A: [], B: []) {
    for i in 1..10 {
      B[i] = A[i] + A[i] * A[i];
    }
  }

  var A: [1..10] int;
  A = 3;
  someProc(A, A);
  writeln(A)
  ```

- How do you determine if an array is being modified inside the loop? Do you just check for some `A[i]` being the lhs of a loop? Does it break if you have something like:

  ```chapel
    for i in 1..10 {
      B[i] = A[i] + A[i] * A[i];
      A = i;
      C[i] = A[i] + A[i] * A[i];
    }
  ```


- Does this have any performance impact on lcals?

File: compiler/optimizations/replaceArrayAccessesWithRefTemps.cpp [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+      if (!indexMove) {
+        // If we couldn't find an expected index move, skip the optimization
+        if (DEBUG_RAAWRT) {
+          printf("%s:%d: Couldn't find index move.  "
+                 "Not replacing accesses in loop\n",
+                 forLoop->fname(), forLoop->linenum());
+        }
+        continue;
+      }
+      for_vector(BaseAST, astNode, asts) {
```
Comment (ronawho) [inline compiler/optimizations/replaceArrayAccessesWithRefTemps.cpp]: Can this just loop over gContextCallExprs?

File: compiler/main/driver.cpp [hunk 2]
```diff
@@ -652,6 +654,7 @@ static ArgumentDescription arg_desc[] = {
  {"privatization", ' ', NULL, "Enable [disable] privatization of distributed arrays and domains", "n", &fNoPrivatization, "CHPL_DISABLE_PRIVATIZATION", NULL},
  {"remote-value-forwarding", ' ', NULL, "Enable [disable] remote value forwarding", "n", &fNoRemoteValueForwarding, "CHPL_DISABLE_REMOTE_VALUE_FORWARDING", NULL},
  {"remove-copy-calls", ' ', NULL, "Enable [disable] remove copy calls", "n", &fNoRemoveCopyCalls, "CHPL_DISABLE_REMOVE_COPY_CALLS", NULL},
+ {"replace-array-accesses-with-ref-temps", ' ', NULL, "Enable [disable] replacing array accesses with reference temps", "N", &fReplaceArrayAccessesWithRefTemps, NULL, NULL },
```
Comment (ronawho) [inline compiler/main/driver.cpp]: Man page needs to be updated

**Comment by ronawho** [comment]
I still haven't fully digested the actual compiler code, but I'm concerned (maybe unjustly) that the analysis done here isn't enough. 

In particular I'm concerned about the aliasing and interleaving def cases above and similar cases like that and I'd like to see some tests cases added. Other things off the top of my head that might screw this up is aliasing through a `ref` var, aliasing through the array alias operator, array defs from things other than than `A[i] = ...`, and defs from tasks outside the current scope. Things like (and i'm not sure all of these make sense/will compile, but I think you'll get gist):


```chapel
// A/B alias
ref A = B;
for i in 1..10 {
  B[i] = A[i] + A[i] * A[i];
}
```

```chapel
// modification by another task
begin {
  for i in 1..10 by -1 {
    s$.writeEF(true)
    A[i] = i;
  }
}

for i in 1..10 {
  s$;
  A[i] = A[i] * A[i];
  B[i] = A[i] + f(A[i]);
}
```

```chapel
// modification through something other than `this()`
for (a, i) in zip(A, 1..10) {
  A[i] = A[i] * A[i];
  a=i;
  B[i] = A[i] + f(A[i]);
}
```

I think some (most?) of these code snippets should be added as tests. If they pass, then i'd be a lot less concerned. Even if they don't I think I'd still be open to accepting this as an "experimental" optimization and documenting it in [PERFORMANCE.md] (https://github.com/chapel-lang/chapel/blob/master/PERFORMANCE.md)

File: compiler/optimizations/replaceArrayAccessesWithRefTemps.cpp (hunk 1 above)
Comment (daviditen) [inline compiler/optimizations/replaceArrayAccessesWithRefTemps.cpp]: This only wants to look at the ContextCallExprs that are within the forLoop it is looking at, not all ContextCallExprs throughout the program.

File: compiler/main/driver.cpp (hunk 2 above)
Comment (daviditen) [inline compiler/main/driver.cpp]: I updated the man page now.

File: compiler/optimizations/replaceArrayAccessesWithRefTemps.cpp (hunk 1 above)
Comment (ronawho) [inline compiler/optimizations/replaceArrayAccessesWithRefTemps.cpp]: Oh, duh. Don't know how I missed that

File: test/arrays/diten/replaceArrayAccess/zipperModify.chpl [hunk 3]
```diff
@@ -0,0 +1,12 @@
+var A, B: [1..4] int = 1..4;
+
+proc f(i) return i;
+
+for (a, i) in zip(A, 1..4) {
+  A[i] = A[i] * A[i];
```
Comment (ronawho) [inline test/arrays/diten/replaceArrayAccess/zipperModify.chpl]: Does this version still work if you had:

```chapel
B[i] = A[i] + f(A[i]);
a = i;
C[i] = A[i] + f(A[i]);
```

File: compiler/main/driver.cpp (hunk 2 above)
Comment (ronawho) [inline compiler/main/driver.cpp]: Oh, also these seems like a super long name to me, although I admit that I don't have a better one to suggest off the top of my head

File: test/arrays/diten/replaceArrayAccess/zipperModify.chpl (hunk 3 above)
Comment (daviditen) [inline test/arrays/diten/replaceArrayAccess/zipperModify.chpl]: This test doesn't match the pattern this is looking for, so it currently doesn't apply.

File: compiler/main/driver.cpp (hunk 2 above)
Comment (daviditen) [inline compiler/main/driver.cpp]: Yeah, it's one of the longest option names.  I'm open to changing it.

**Comment by ronawho** [comment]
I'm ok with this. I'd prefer a shorter flag name (still can't think of a good one myself), and I think the PR message should be updated to mention that it's not correct for all cases and is going in as an experimental/opt-in optimization.


Other than that I think you just need to add this to PERFORMANCE.md, but that can be done during doc freeze

**Comment by ronawho** [comment]
Oh and maybe the flag help/man page should mention that it's experimental

**Pull request merged**